### PR TITLE
Fix text record resolution and enhance CCIP-Read handling

### PR DESCRIPTION
**Pull request by Douglasacost**
This pull request introduces a new signed-gateway UniversalResolver integration for ENS offchain resolution, adding a `/resolve` endpoint that supports CCIP-Read (ERC-3668) with EIP-712 signatures, and refactors deployment scripts and configuration to support the new model. The changes include new resolver logic, EIP-712 signing, configuration wiring, and a simplified deployment script.

**New ENS Gateway Resolution Functionality:**

- Added `/resolve` route that implements the CCIP-Read (ERC-3668) callback endpoint, decoding ENS resolution requests, routing them to the correct L2 NameService, and returning EIP-712 signed responses for use with the L1 UniversalResolver. [[1]](diffhunk://#diff-750ebcff3ec02953b358dbda9546cf08c9da6b534096567d77d0aaea970a8d54R1-R147) [[2]](diffhunk://#diff-0441f3ab88dfb5720a1b2b4b1bcccff30690c0f573cf95eccf4ed416d8b2a6eaR47-R52) [[3]](diffhunk://#diff-0441f3ab88dfb5720a1b2b4b1bcccff30690c0f573cf95eccf4ed416d8b2a6eaR518)
- Implemented `resolveFromL2` utility to parse DNS-encoded ENS names, support `addr`, `addr-multichain`, and `text` lookups, and return ABI-encoded results.
- Added `signResolutionResponse` for EIP-712 signing of resolution results, matching the L1 UniversalResolver contract.

**Configuration and Setup Updates:**

- Extended `setup.ts` to wire up the new resolver signer, L1 resolver address, and signature TTL, and export them for use in the new route. [[1]](diffhunk://#diff-1ee4e2e3fc4a59ca4d25cfcb449ecef2dfc2a7108e4dd5218dd575941da372ecL1-R83) [[2]](diffhunk://#diff-1ee4e2e3fc4a59ca4d25cfcb449ecef2dfc2a7108e4dd5218dd575941da372ecL76-R119) [[3]](diffhunk://#diff-1ee4e2e3fc4a59ca4d25cfcb449ecef2dfc2a7108e4dd5218dd575941da372ecL105-R142) [[4]](diffhunk://#diff-1ee4e2e3fc4a59ca4d25cfcb449ecef2dfc2a7108e4dd5218dd575941da372ecL128-R183)
- Updated the ENS NameService interface to include the `getTextRecord` function needed by the gateway.

**Deployment Script Changes:**

- Refactored the L1 deployment script to support the new signed-gateway UniversalResolver, removing the StorageProofVerifier/SparseMerkleTree logic and adding support for setting the ENS resolver in a single broadcast if desired. [[1]](diffhunk://#diff-d9e85f7d19dbcfb84027028f9de0c9edcb6e3ef048c09a18fe29ceda10e77709L5-L9) [[2]](diffhunk://#diff-d9e85f7d19dbcfb84027028f9de0c9edcb6e3ef048c09a18fe29ceda10e77709L22-R36) [[3]](diffhunk://#diff-d9e85f7d19dbcfb84027028f9de0c9edcb6e3ef048c09a18fe29ceda10e77709R45-R49)

These changes collectively enable secure, offchain ENS resolution via a gateway that signs responses for L1 verification, modernizing the resolver infrastructure and deployment flow.

### CODE REVIEW

File: src/nameservice/UniversalResolver.sol [hunk 1]
```diff
@@ -1,72 +1,134 @@
 // SPDX-License-Identifier: BSD-3-Clause-Clear
 
 /**
- * @title UniversalResolver for resolving ens subdomains based on names registered on L2
- * @dev This contract is based on ClaveResolver that can be found in this repository:
- * https://github.com/getclave/zksync-storage-proofs
+ * @title UniversalResolver
```
Comment (aliXsed) [inline src/nameservice/UniversalResolver.sol]: Please create a new contract instead of changing the previous UniversalResolver. The code in the current contract both represent what has gone on chain at deployed addresses, and also provide the base for fixing what started to fail following the ZKSync upgrade.

**Review by aliXsed — COMMENTED** [review]
## Consolidated Review — PR #111

The changeset introduces a well-structured CCIP-Read signed-gateway flow. Below are findings ordered by severity, combining my earlier inline comments with additional observations.

---

### 1. [Solidity] Create a new contract — don't modify the existing `UniversalResolver`

The deployed contract's source should be preserved as a historical artifact. The new contract has a fundamentally different verification model (EIP-712 signatures vs. storage proofs). It should live in a separate file/contract. *(Reiterating my earlier comment.)*

### 2. [Solidity] `Ownable` → `Ownable2Step`

Per the project's own standards: *"When a contract requires an owner, prefer `Ownable2Step` over `Ownable`."* This contract requires ownership for `setUrl` and `setTrustedSigner`. A single-step `transferOwnership` to a wrong address is irrecoverable and bricks signer management.

### 3. [Solidity] `setUrl` missing empty-URL validation

The constructor validates `bytes(_url).length == 0 → revert EmptyUrl()`, but `setUrl` does not. The owner can accidentally brick all subdomain resolution. *(Reiterating my earlier inline suggestion — CEI ordering + empty check.)*

### 4. [Solidity] Split `setTrustedSigner` into `trustSigner` / `revokeSigner`

`setTrustedSigner(addr, false)` is semantically ambiguous — "set trusted signer" that actually removes trust. Two distinct functions simplify access-control auditing and event decoding. *(Reiterating my earlier comment.)*

### 5. [Solidity] Add `_data.length >= 4` check in `resolve()`

Short calldata causes a panic (out-of-bounds) rather than a clean `CallDataTooShort` revert. The custom error already exists but the guard is missing. *(Copilot flagged this — verify it's addressed.)*

### 6. [Solidity] Fix `addr(bytes32,uint256)` bare-domain return type

Bare-domain handler returns `abi.encode(address(0))` for multichain, but ENSIP-11 specifies `bytes` return. Should return `abi.encode(bytes(""))`. *(Copilot flagged this — verify it's addressed.)*

### 7. [Gateway] `parseResolutionSignatureTtl` style — convert to IIFE

Every other derived config in `setup.ts` is an inline `const`. The 23-line named function stands out. The validation is necessary (prevents emitting signatures guaranteed to revert on-chain), but it should be an IIFE to match the file's style:

```ts
const resolutionSignatureTtlSeconds = (() => {
  const raw = process.env.RESOLUTION_SIGNATURE_TTL_SECONDS;
  if (raw === undefined || raw === "") return 60;
  const parsed = Number(raw);
  if (!Number.isFinite(parsed) || !Number.isInteger(parsed))
    throw new Error(`Invalid RESOLUTION_SIGNATURE_TTL_SECONDS: "${raw}" is not a finite integer`);
  if (parsed <= 0)
    throw new Error(`RESOLUTION_SIGNATURE_TTL_SECONDS must be > 0, got ${parsed}`);
  if (parsed > MAX_RESOLUTION_SIGNATURE_TTL_SECONDS)
    throw new Error(`RESOLUTION_SIGNATURE_TTL_SECONDS must be <= ${MAX_RESOLUTION_SIGNATURE_TTL_SECONDS}, got ${parsed}`);
  return parsed;
})();
```

### 8. [Gateway] Unused `hexlify` import in `resolve.ts`

`hexlify` is imported from ethers on line 1 but never used. Remove it.

### 9. [Gateway] DNS parsing — validate TLD against `parentTLD`

The gateway uses `parsed.sub` and `parsed.domain` for routing but never validates `parsed.tld`. A sanity check against the configured `parentTLD` would reject obviously malformed names early. Low priority since the L1 resolver already constrains which names reach the gateway.

### 10. [Tests] Missing fuzz tests for TTL boundaries

Per project standards (`testFuzz_` required for arithmetic/functional logic): TTL boundary checks and DNS parsing have no fuzz coverage. At minimum `testFuzz_ResolveWithSig_ExpiresAt` for various timestamp values.

---

### Items I reviewed but am NOT flagging:

- **`/resolveL2` vs `/resolve` redundancy** — They serve different purposes (simple HTTP API vs. CCIP-Read callback). Core L2 logic is already in `resolveFromL2.ts`. No shared-function refactor needed.
- **Route path `POST /` inside the resolve router** — Standard Express sub-router pattern, consistent with `names.ts`. The mount path (`/resolve`) *is* the path.
- **Error handling inconsistency** — The new code consistently uses `HttpError` + `asyncHandler`. The older endpoints in `index.ts` use manual try/catch — that predates this PR and isn't introduced by it.

### Action summary (priority order):

| # | File | Action |
|---|------|--------|
| 1 | `UniversalResolver.sol` | New contract file, don't modify existing |
| 2 | `UniversalResolver.sol` | `Ownable2Step` |
| 3 | `UniversalResolver.sol` | Empty-URL check in `setUrl` |
| 4 | `UniversalResolver.sol` | `trustSigner` / `revokeSigner` instead of `setTrustedSigner(bool)` |
| 5 | `UniversalResolver.sol` | `_data.length >= 4` guard |
| 6 | `UniversalResolver.sol` | Fix multichain bare-domain return type |
| 7 | `setup.ts` | IIFE for TTL validation |
| 8 | `resolve.ts` | Remove unused `hexlify` import |
| 9 | `resolve.ts` | Optional: validate `parsed.tld` |
| 10 | `UniversalResolver.t.sol` | Add fuzz tests |

File: src/nameservice/UniversalResolver.sol [hunk 2]
```diff
@@ -1,72 +1,134 @@
...
+import {EIP712} from "@openzeppelin/contracts/utils/cryptography/EIP712.sol";
 
 /// @title IExtendedResolver
 /// @notice ENSIP-10: Wildcard Resolution
 interface IExtendedResolver {
     function resolve(bytes calldata name, bytes calldata data) external view returns (bytes memory);
 }
 
-contract UniversalResolver is IExtendedResolver, IERC165, Ownable {
+contract UniversalResolver is IExtendedResolver, IERC165, Ownable, EIP712 {
```
Comment (aliXsed) [inline src/nameservice/UniversalResolver.sol:28]: ### [1/10] Create a new contract — don't modify the existing `UniversalResolver`

The deployed contract's source should be preserved as a historical artifact. The new contract has a fundamentally different verification model (EIP-712 signatures vs. storage proofs). It should live in a separate file/contract. *(Reiterating my earlier comment.)*

**Review by aliXsed — COMMENTED** [review]
Consolidated inline review — 10 items across the changeset, ordered by priority.

File: src/nameservice/UniversalResolver.sol [hunk 3]
```diff
@@ -1,72 +1,134 @@
...
+ *      and accepts the response only if it matches a registered trusted signer.
+ *
+ *      This replaces the earlier zkSync storage-proof design which depended on
+ *      per-batch state roots being committed to L1 — that path was broken when
+ *      zkSync Era migrated settlement to ZK Gateway (~July 30, 2025).
  */
 pragma solidity ^0.8.26;
 
 import {IERC165} from "lib/forge-std/src/interfaces/IERC165.sol";
 import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
```
Comment (aliXsed) [inline src/nameservice/UniversalResolver.sol:18]: ### [2/10] `Ownable` → `Ownable2Step`

Per the project's own standards: *"When a contract requires an owner, prefer `Ownable2Step` over `Ownable`."* This contract requires ownership for `setUrl` and `setTrustedSigner`. A single-step `transferOwnership` to a wrong address is irrecoverable and bricks signer management.

File: src/nameservice/UniversalResolver.sol [hunk 4]
```diff
@@ -1,72 +1,134 @@
...
-        textRecordsSlot = 9;
+        isTrustedSigner[_initialSigner] = true;
+        trustedSignerCount = 1;
+        emit TrustedSignerUpdated(_initialSigner, true);
     }
 
+    /// @notice Update the CCIP-Read gateway URL.
     function setUrl(string memory _url) external onlyOwner {
+        emit UrlUpdated(url, _url);
         url = _url;
```
Comment (aliXsed) [inline src/nameservice/UniversalResolver.sol:98]: ### [3/10] `setUrl` missing empty-URL validation

The constructor validates `bytes(_url).length == 0 → revert EmptyUrl()`, but `setUrl` does not. The owner can accidentally brick all subdomain resolution by setting an empty URL.

*(Reiterating my earlier inline suggestion — add CEI ordering + empty check.)*

```suggestion
    function setUrl(string memory _url) external onlyOwner {
        bytes memory b = bytes(_url);
        if (b.length == 0) revert EmptyUrl();

        string memory oldUrl = url;
        url = _url;
        emit UrlUpdated(oldUrl, url);
    }
```

File: src/nameservice/UniversalResolver.sol [hunk 5]
```diff
@@ -1,72 +1,134 @@
...
     }
 
+    /// @notice Enable or disable a trusted gateway signer.
+    /// @dev Keeps `trustedSignerCount` in sync and enforces a floor of 1 so the
+    ///      owner cannot brick resolution by disabling the last signer.
+    function setTrustedSigner(address signer, bool trusted) external onlyOwner {
```
Comment (aliXsed) [inline src/nameservice/UniversalResolver.sol:104]: ### [4/10] Split `setTrustedSigner` into `trustSigner` / `revokeSigner`

`setTrustedSigner(addr, false)` is semantically ambiguous — "set trusted signer" that actually removes trust. Two distinct functions simplify access-control auditing and event decoding. *(Reiterating my earlier comment.)*

File: src/nameservice/UniversalResolver.sol [hunk 6]
```diff
@@ -105,87 +167,98 @@ contract UniversalResolver is IExtendedResolver, IERC165, Ownable {
...
-        if (bytes(sub).length == 0) {
-            return abi.encodePacked(domainOwner);
+        // Explicit length check so short calldata reverts with a controlled error
+        // instead of a panic on the slice below.
+        if (_data.length < 4) {
+            revert CallDataTooShort(_data.length);
         }
 
+        // Dispatch only on supported selectors so the gateway is never asked for nonsense.
         bytes4 functionSelector = bytes4(_data[:4]);
```
Comment (aliXsed) [inline src/nameservice/UniversalResolver.sol:183]: ### [5/10] Add `_data.length >= 4` check in `resolve()`

Short calldata causes a panic (out-of-bounds) rather than a clean `CallDataTooShort` revert. The custom error `CallDataTooShort` already exists but the guard is missing before this slice. *(Copilot flagged this — verify it's addressed.)*

```suggestion
        if (_data.length < 4) revert CallDataTooShort();
        bytes4 functionSelector = bytes4(_data[:4]);
```

File: src/nameservice/UniversalResolver.sol [hunk 7]
```diff
@@ -105,87 +167,98 @@ contract UniversalResolver is IExtendedResolver, IERC165, Ownable {
...
+        // the ENS "no record" convention: zero address for addr queries, empty string
+        // for text queries. The resolver only exists to answer subdomain lookups — it
+        // holds no state about the parent name. If a specific address needs to be
+        // associated with the bare domain, set it via a different resolver at the
+        // ENS registry level.
+        if (bytes(sub).length == 0) {
+            if (functionSelector == _TEXT_SELECTOR) {
+                return abi.encode("");
+            }
             if (functionSelector == _ADDR_MULTICHAIN_SELECTOR) {
```
Comment (aliXsed) [inline src/nameservice/UniversalResolver.sol:207]: ### [6/10] Fix `addr(bytes32,uint256)` bare-domain return type

Bare-domain handler returns `abi.encode(address(0))` for multichain, but ENSIP-11 defines `addr(bytes32,uint256)` as returning `bytes`. Should return `abi.encode(bytes(""))` for the multichain case. *(Copilot flagged this — verify it's addressed.)*

File: clk-gateway/src/setup.ts [hunk 8]
```diff
@@ -1,63 +1,114 @@
...
+  if (parsed > MAX_RESOLUTION_SIGNATURE_TTL_SECONDS) {
+    throw new Error(
+      `Invalid RESOLUTION_SIGNATURE_TTL_SECONDS: must be <= ${MAX_RESOLUTION_SIGNATURE_TTL_SECONDS} ` +
+        `(L1 resolver _MAX_SIGNATURE_TTL), got ${parsed}`,
+    );
+  }
+  return parsed;
+}
+
+const resolutionSignatureTtlSeconds = parseResolutionSignatureTtl(
```
Comment (aliXsed) [inline clk-gateway/src/setup.ts:105]: ### [7/10] Convert `parseResolutionSignatureTtl` to IIFE for consistency

Every other derived config in this file is an inline `const` expression. The 23-line named function stands out. The validation is necessary (prevents emitting signatures guaranteed to revert on-chain), but converting to an IIFE would match the file's style:

```ts
const resolutionSignatureTtlSeconds = (() => {
  const raw = process.env.RESOLUTION_SIGNATURE_TTL_SECONDS;
  if (raw === undefined || raw === "") return 60;
  const parsed = Number(raw);
  if (!Number.isFinite(parsed) || !Number.isInteger(parsed))
    throw new Error(`Invalid RESOLUTION_SIGNATURE_TTL_SECONDS: "${raw}" is not a finite integer`);
  if (parsed <= 0)
    throw new Error(`RESOLUTION_SIGNATURE_TTL_SECONDS must be > 0, got ${parsed}`);
  if (parsed > MAX_RESOLUTION_SIGNATURE_TTL_SECONDS)
    throw new Error(`RESOLUTION_SIGNATURE_TTL_SECONDS must be <= ${MAX_RESOLUTION_SIGNATURE_TTL_SECONDS}, got ${parsed}`);
  return parsed;
})();
```

File: clk-gateway/src/routes/resolve.ts [hunk 9]
```diff
@@ -0,0 +1,163 @@
+import { AbiCoder, dataSlice, getAddress, hexlify, isAddress, isHexString } from "ethers"
```
Comment (aliXsed) [inline clk-gateway/src/routes/resolve.ts:1]: ### [8/10] Unused `hexlify` import

`hexlify` is imported here but never used anywhere in this file. Remove it.

File: clk-gateway/src/routes/resolve.ts [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+      decodedName = n
+      decodedData = d
+    } catch (_err: unknown) {
+      payload = dataSlice(ccipCallData, 4)
+      const [n, d] = abi.decode(["bytes", "bytes"], payload)
+      decodedName = n
+      decodedData = d
+    }
+
+    const parsed = parseDnsDomain(Buffer.from(decodedName.slice(2), "hex"))
```
Comment (aliXsed) [inline clk-gateway/src/routes/resolve.ts:111]: ### [9/10] Consider validating `parsed.tld` against `parentTLD`

The gateway uses `parsed.sub` and `parsed.domain` for routing but never validates `parsed.tld`. A sanity check against the configured `parentTLD` would reject obviously malformed names early. Low priority since the L1 resolver already constrains which names reach the gateway.

File: test/nameservice/UniversalResolver.t.sol [hunk 11]
```diff
@@ -0,0 +1,404 @@
...
+        assertFalse(resolver.supportsInterface(0xdeadbeef));
+    }
+
+    // --- sanity: initial signer was set ---
+
+    function test_InitialSignerIsTrusted() public view {
+        assertTrue(resolver.isTrustedSigner(signer));
+        assertFalse(resolver.isTrustedSigner(backupSigner));
+    }
+}
```
Comment (aliXsed) [inline test/nameservice/UniversalResolver.t.sol:404]: ### [10/10] Missing fuzz tests for TTL boundaries

Per project standards (`testFuzz_` required for arithmetic/functional logic): TTL boundary checks and DNS parsing have no fuzz coverage. At minimum `testFuzz_ResolveWithSig_ExpiresAt` for various timestamp values would strengthen confidence in the TTL/expiry arithmetic.

File: src/nameservice/UniversalResolver.sol [hunk 12]
```diff
@@ -105,87 +167,98 @@ contract UniversalResolver is IExtendedResolver, IERC165, Ownable {
...
-    /// @param _name The name to resolve which must be a pack of length prefixed names for subdomain, domain and top.
-    /// example: b"\x07example\x05clave\x03eth"
-    ///
-    /// @param _data The ABI encoded data for the underlying resolution function (Eg, addr(bytes32), text(bytes32,string), etc).
+    /// @notice ENSIP-10 entry point. Triggers CCIP-Read lookup via OffchainLookup revert.
+    /// @param _name DNS-encoded name (e.g. b"\x07example\x05clave\x03eth")
+    /// @param _data ABI-encoded ENS resolution call (addr / addr-multichain / text)
     function resolve(bytes calldata _name, bytes calldata _data) external view returns (bytes memory) {
-        (string memory sub, string memory dom,) = _parseDnsDomain(_name);
+        (string memory sub,,) = _parseDnsDomain(_name);
```
Comment (aliXsed) [inline src/nameservice/UniversalResolver.sol:174]: ### [Addendum] `dom` is discarded — consider using it for validation

`resolve()` extracts `(sub,,)` and discards `dom` entirely. The contract never validates that the domain portion of the DNS name matches its intended domain.

**Impact:** If the ENS registry mistakenly points a different domain at this resolver, it would happily trigger `OffchainLookup` for that unrelated domain with no guard. The gateway compensates (it routes by `parsed.domain`), but the contract itself is blind to it.

**Suggested fix:** At minimum, pass `dom` through to the gateway in the `callData`/`extraData` so it can be validated end-to-end. Better yet, store the expected domain in the contract and reject mismatches on-chain — the gas cost is trivial for a `view` function.

```solidity
// Instead of:
(string memory sub,,) = _parseDnsDomain(_name);

// Consider:
(string memory sub, string memory dom,) = _parseDnsDomain(_name);
// ... use dom for validation or pass it through
```

**Comment by aliXsed** [comment]
Tracking issue for evaluating a trustless alternative to the signed-gateway interim solution: **NodleCode/meta#169** — *Evaluate Unruggable Gateways ZkSync support as replacement for trusted signed-gateway.*

The signed-gateway model in this PR is standards-compliant (ERC-3668, ENSIP-10) and matches what Coinbase/Uniswap/ENS reference use in production. However, [Unruggable Gateways](https://github.com/unruggable-labs/unruggable-gateways) already has ZkSync support (`ZKSyncRollup` + `ZKSyncProver` + `ZKSyncVerifierHooks`) and could potentially restore the trustless verification model we had before the ZkSync proof format break.

File: src/nameservice/UniversalResolver.sol (hunk 7 above)
Comment (Douglasacost) [reply src/nameservice/UniversalResolver.sol:145]: Already addressed. The bare-domain path branches on _ADDR_MULTICHAIN_SELECTOR and returns abi.encode(bytes("")) per ENSIP-11 (line 211–214). The regular addr(bytes32) path falls through to abi.encode(address(0)). The gateway side (resolveFromL2.ts) also encodes multichain results as bytes (not address) in both the found and not-found cases. 
Covered by test_Resolve_BareDomain_AddrMultichain_ReturnsEmptyBytes and test_ResolveWithSig_AddrMultichain_EmptyRecord_HappyPath. Updated the inline comment to document all three return types explicitly. 
 
See commits: 1243a8c (original fix), 647df3d (comment clarification).

File: clk-gateway/src/setup.ts (hunk 8 above)
Comment (Douglasacost) [reply clk-gateway/src/setup.ts]: Applied

**Review by aliXsed — APPROVED** [review]
